### PR TITLE
#M05 Creating Hackathon Database Models

### PR DESCRIPTION
**Pull request by Ri-Dearg**
These models are a draft that will fix issue #61 
**TLDR** at the end.

**Overview**
The models.py file contains the models built from the schema [here](https://raw.githubusercontent.com/Code-Institute-Community/ci-hackathon-app/master/static/img/documentation/hackathon-ERD.png) on the [Development Wiki](https://github.com/Code-Institute-Community/ci-hackathon-app/wiki/Development). I have followed the schema to the best of my ability. I definitely need other contributors to take a look at the fields for the Foreign Keys because I did change some of them.

I have given my reasoning for the choices in comments above each Foreign Key field. If anybody can see any errors or has a better idea of how to relate them, the advice is well appreciated! :-)

**Testing**
I have created each model in my own environment, it all seems grand. I deleted the DB and remigrated everything, I haven't had any errors yet. Models have a testing file that checks they can be created correctly and their `str` method.

**To Do**
- Create a global method to automatically fill the created_by field on creation. Maybe find a way to prevent it from being edited.
- Beautify the admin pages for the table relations and models. Some models depend on other models being present, so it can be really unintuitive to use the admin page to create them because you have to jump around.
Displaying info from the related tables and making it editable on a different model's page would make this way easier! However, I think it could be started as a new issue. The models are needed to start work on other issues and it is a job in itself.

**Possible Bugs**
- For the HackTeam model, it uses a ManyToMany field for "participants". I thought this was useful in case a User was part of a few Hackathons. However, it means that a user could be part of more than one team for the same hackathon. I think this could be solved with a custom save method.

I know it is a long read, so thank you again for any help!

**TLDR**:
Please take a look at the Foreign Key fields and their comments and let me know if you see any issues. I'll fix them up and make the commits. ;-)

### CODE REVIEW

File: hackathon/models.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    """Model representing a Hackathon. It is connected by a foreign key to 
+    User, HackAwards and HackTeam. Optional Fields: judges, organiser.
+    "awards" and "teams" are related tables. They have been moved to
+    HackAwardCategory and HackTeam respectively. Please see comments there."""
+    created = models.DateTimeField(default=timezone.now)
+    updated = models.DateTimeField(auto_now=True)
+    # Each model can only be created by one user: One To Many
+    created_by = models.ForeignKey(User,
+                                   on_delete=models.CASCADE,
+                                   related_name="hackathon_created_by")
```
Comment (Ri-Dearg) [inline hackathon/models.py:29]: comment test

**Review by stefdworschak — CHANGES_REQUESTED** [review]
Overall a really good Pull Request. Some suggestions on how to proceed and let me know if you have any questions.

File: hackathon/admin.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+                     HackProjectScoreCategory)
+
+
+# Register your models here.
+admin.site.register(Hackathon)
+admin.site.register(HackAwardCategory)
+admin.site.register(HackTeam)
+admin.site.register(HackProject)
+admin.site.register(HackProjectScore)
+admin.site.register(HackProjectScoreCategory)
```
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply hackathon/admin.py:16]: I think that's just GitHub cutting it out. They're in in the files.

File: hackathon/models.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+# My reasoning is explained in comments above the ForeignKey fields.
+
+
+# Create your models here.
+class Hackathon(models.Model):
+    """Model representing a Hackathon. It is connected by a foreign key to 
+    User, HackAwards and HackTeam. Optional Fields: judges, organiser.
+    "awards" and "teams" are related tables. They have been moved to
+    HackAwardCategory and HackTeam respectively. Please see comments there."""
+    created = models.DateTimeField(default=timezone.now)
```
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply hackathon/models.py:24]: Yeah, I thought that as well. In the Django tutorial for Boutique Ado this is used because it is editable afterwards. I wasn't sure if it was necessary, so I'll swap it out.

File: hackathon/models.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+# The "updated" field isn't editable in admin. Neither is "submission" field.
+# A custom method to auto-fill created_by fields can be made
+
+# When it seemed relevant, I moved the ForeignKey field to another model and
+# used the one listed in the schema as the related_name. I may have been
+# erroneous to do so but it can be corrected easily.
+# My reasoning is explained in comments above the ForeignKey fields.
+
+
+# Create your models here.
```
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply hackathon/models.py:18]: Deleted

File: hackathon/models.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+class HackAwardCategory(models.Model):
+    """Model representing a HackAwardCategory. It is connected by a foreign key to 
+    User, Hackathon and HackProject. Optional fields: winning_project."""
+    created = models.DateTimeField(default=timezone.now)
+    updated = models.DateTimeField(auto_now=True)
+    # Each model can only be created by one user: One To Many
+    created_by = models.ForeignKey(User,
+                                   on_delete=models.CASCADE,
+                                   related_name="hackawardcategory_created_by")
```
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply hackathon/models.py:58]: Cool, done.

File: hackathon/models.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    display_name = models.CharField(default="", max_length=254)
+    description = models.TextField()
+    # a Category will only apply to one Hackathon and 
+    # a Hackathon has numerous categories: One to Many.
+    # If the category was going to be reused, instead, use Many to Many.
+    hackathon = models.ForeignKey("Hackathon",
+                                  on_delete=models.CASCADE,
+                                  related_name="awards")
+    # One category can have one winner: One to One
+    winning_project = models.OneToOneField("HackProject",
```
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply hackathon/models.py:68]: You use quotes to reference an object which has not yet been defined. From here:
https://docs.djangoproject.com/en/3.1/ref/models/fields/#foreignkey

> If you need to create a relationship on a model that has not yet been defined, you can use the name of the model, rather than the model object itself

**Review by Ri-Dearg — COMMENTED** [review]
I implemented the requested changes. Main points:
- I'm using `default=""` because Django docs recommend it for string fields. If you prefer, I'll change it.
- I removed the quotes around "Hackathon" but left them for "HackProject", take a look at the comment.
- I think I made that last change for the model correctly, but let me know if it's right!

Thank you for all the effort ye put into supporting the project, it's great experience for me. :-)

File: .theia/settings.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply .theia/settings.json]: Done! Added it to .gitignore. I hadn't realised it was added.

File: hackathon/admin.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/admin.py:16]: Cool.

File: hackathon/models.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    User, HackAwards and HackTeam. Optional Fields: judges, organiser.
+    "awards" and "teams" are related tables. They have been moved to
+    HackAwardCategory and HackTeam respectively. Please see comments there."""
+    created = models.DateTimeField(default=timezone.now)
+    updated = models.DateTimeField(auto_now=True)
+    # Each model can only be created by one user: One To Many
+    created_by = models.ForeignKey(User,
+                                   on_delete=models.CASCADE,
+                                   related_name="hackathon_created_by")
+    display_name = models.CharField(default="", max_length=254)
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/models.py:27]: Okay, just learnt something new. Did not know that, but let's follow the Django documentation then.

File: hackathon/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/models.py:65]: Yep. Let's leave that then.

File: hackathon/models.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    created_by = models.ForeignKey(User,
+                                   on_delete=models.CASCADE,
+                                   related_name="hackprojectscore_created_by")
+    # One Judge can give one score - One to One
+    judge = models.OneToOneField(User, on_delete=models.CASCADE)
+    # One score is for one project, a project has numerous scores: One to Many
+    project = models.ForeignKey("HackProject",
+                                on_delete=models.CASCADE,
+                                related_name="scores")
+    # A score applies to one category, a category has many scores: One to Many
```
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/models.py:151]: The change looks great! Thanks for that.

File: hackathon/models.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    """Model representing a Hackathon. It is connected by a foreign key to 
+    User, HackAwards and HackTeam. Optional Fields: judges, organiser.
+    "awards" and "teams" are related tables. They have been moved to
+    HackAwardCategory and HackTeam respectively. Please see comments there."""
+    created = models.DateTimeField(auto_now_add=True)
+    updated = models.DateTimeField(auto_now=True)
+    # Each model can only be created by one user: One To Many
+    created_by = models.ForeignKey(User,
+                                   on_delete=models.CASCADE,
+                                   related_name="hackathon_created_by")
```
Comment (stefdworschak) [inline hackathon/models.py:26]: Sorry, I should have mentioned this in the other comment, but I would just rename all the `related_name`s for  `created_by` to be just the plural of the model name in lower case. I think that would be the best seeing as it they are basically a collection of that model created by a user.

**Comment by stefdworschak** [comment]
Also, don't forget to fix the merge conflict.

File: hackathon/models.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Ri-Dearg) [reply hackathon/models.py:27]: Actually, I was doing the same thing myself until recently, then I found this. It only has a few rules, but it's useful:
https://github.com/rocioar/flake8-django

File: hackathon/models.py (hunk 10 above)
Comment (Ri-Dearg) [reply hackathon/models.py:26]: Okay, makes sense, I'll do that now.

**Comment by Ri-Dearg** [comment]
Okay, I changed the related names to all plural, I merged the conflict in the .gitignore adding both lines. And there are some basic str method tests in there too now.

I think it should be good to go. If you have any other corrections, let me know and I'll get right to it! :-)

File: hackathon/models.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (stefdworschak) [reply hackathon/models.py:27]: That's really useful!

**Review by stefdworschak — APPROVED** [review]
LGTM

Really good pull request! :+1:
